File: mycipher.py
#encoding: UTF-8

# Python Class for AES encryption
"""
    https://github.com/arajapandi/php-python-encrypt-decrypt
    Example Usage
    enc_str = cipher.encrypt('secret')
    enc_str = cipher.decrypt(enc_str)
    print enc_str; #secret
"""

from Crypto.Cipher import AES
from Crypto import Random
import base64
import hashlib
import sys


class MyCipher:
    #Default Key for encryption
    rawkey      = 'asdfa923aksadsYahoasdw998sdsads'
    method      = AES.MODE_CFB
    blocksize   = 32  # 16, 32..etc
    padwith     = '`' # padding value for string  
    
    #lambda function for padding
    pad         = lambda self, s: s + (self.blocksize - len(s) % self.blocksize) * self.padwith
    
    """
    construct for cipher class - get, set key and iv
    """
    def __init__(self, iv='', key='', hash_iv=True, hash_key=True):
        
        self.input_iv = iv
        self.input_key = key

        if not key:
            key = self.rawkey

        if hash_key:
            self.key = (
                hashlib.sha256(key.encode('utf-8'))
                .hexdigest()[:32]
                .encode('utf-8')
            )

        if not iv:
            self.iv = Random.get_random_bytes(16)
        else:
            self.iv = iv.encode('utf-8')

        if hash_iv:
            self.iv = hashlib.sha256(self.iv).hexdigest()[:16].encode('utf-8')

    def get_cipher(self):
        return AES.new(self.key,
                       self.method,
                       iv=self.iv,
                       segment_size=128)
    
    """
    Encrypt given string using AES encryption standard
    """
    def encrypt(self, raw):
        cipher = self.get_cipher()
        return base64.b64encode(cipher.encrypt(self.pad(raw).encode('utf-8')))
    
    """
    Decrypt given string using AES standard
    """
    def decrypt(self, encrypted):
        encrypted = base64.b64decode(encrypted)
        cipher = self.get_cipher()
        return cipher.decrypt(encrypted).rstrip(self.padwith.encode('utf-8'))

    '''
    Encrypt given string using AES encryption standard, but prepends
    the iv and then encodes the composite bytes object
    '''
    def encrypt_includes_iv(self, raw):

        cipher = self.get_cipher()
        out = base64.b64encode(self.iv +
                               cipher.encrypt(self.pad(raw).encode('utf-8')))
        return out

    '''
    Decrypt string using AES encryption standard, but expects the first 16
    bytes to be the iv_hash
    '''
    def decrypt_includes_iv(self, encrypted):
        encrypted = base64.b64decode(encrypted)
        iv_hash, encrypted = encrypted[:16], encrypted[16:]

        cipher = AES.new(self.key,
                         self.method,
                         iv_hash,
                         segment_size=128)

        return cipher.decrypt(encrypted).rstrip(self.padwith.encode('utf-8'))

File: mycipher.php
<?php
/**
 * Valid encryption methods AES-256-CFB 
 * 
 * $cypher = new MyCypher($iv);
 * $php_encrypted      = $cypher->encrypt('test');
 * $php_decrypted      = $cypher->decrypt($php_encrypted);
 */
class MyCipher {

    private $key = 'asdfa923aksadsYahoasdw998sdsads';
    private $iv = null;
    private $method = "AES-256-CFB";
    private $blocksize = 32;
    private $padwith = '`';

    /*
     * construct for cipher class - get, set key and iv
     */

    function __construct($iv = null, $key = null) {

        $this->input_key = $key;

        if (!empty($key)) {
            $this->key = $key;
        }

        $this->key = substr(hash('sha256', $this->key), 0, 32);

        $this->input_iv = $iv;

        if (!empty($iv)) {
            $this->iv = $iv;
        } else {
            $this->iv = openssl_random_pseudo_bytes(16);
        }

        $this->iv = substr(hash('sha256', $this->iv), 0, 16);
    }

    /*
     * Encrypt given string using AES encryption standard
     */

    public function encrypt($secret) {

        try {

            $padded_secret = $secret . str_repeat($this->padwith, ($this->blocksize - strlen($secret) % $this->blocksize));
            $encrypted_string = openssl_encrypt($padded_secret, $this->method, $this->key, OPENSSL_RAW_DATA, $this->iv);
            $encrypted_secret = base64_encode($encrypted_string);
            return $encrypted_secret;
        } catch (Exception $e) {
            die('Error : ' . $e->getMessage());
        }
    }

    /*
     * Decrypt given string using AES standard
     */

    public function decrypt($secret) {
        try {
            $decoded_secret = base64_decode($secret);
            $decrypted_secret = openssl_decrypt($decoded_secret, $this->method, $this->key, OPENSSL_RAW_DATA, $this->iv);
            return rtrim($decrypted_secret, $this->padwith);
        } catch (Exception $e) {
            die('Error : ' . $e->getMessage());
        }
    }
    

    public function encrypt_includes_iv($secret) {

        try {

            $padded_secret = $secret . str_repeat($this->padwith, ($this->blocksize - strlen($secret) % $this->blocksize));
            $encrypted_string = openssl_encrypt($padded_secret, $this->method, $this->key, OPENSSL_RAW_DATA, $this->iv);
            $encrypted_secret = base64_encode($this->iv . $encrypted_string);

            return $encrypted_secret;
        } catch (Exception $e) {
            die('Error : ' . $e->getMessage());
        }
    }

    public function decrypt_includes_iv($secret) {
        try {
            $decoded_secret = base64_decode($secret);
            $iv_hash = substr($decoded_secret, 0, 16);
            $decoded_secret = substr($decoded_secret, 16);
            $decrypted_secret = openssl_decrypt($decoded_secret, $this->method, $this->key, OPENSSL_RAW_DATA, $iv_hash);
            return rtrim($decrypted_secret, $this->padwith);
        } catch (Exception $e) {
            die('Error : ' . $e->getMessage());
        }
    }
}
